### PR TITLE
LLM Question Generation: Revisions/Saving

### PR DESCRIPTION
**Pull request by jlevine18**
More copying from the private repo, allows for question revision in the ai-generate page, as well as saving generations in course repos.

### CODE REVIEW

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 1]
```diff
@@ -3,20 +3,104 @@ import { OpenAI } from 'openai';
...
+        
+        Here is some context that may help you respond to the user. This context may include example questions, documentation, or other information that may be helpful.
+        
+        ${context}
+        
+        `;
+}
+
+/**
+ * Builds the context string, consisting of relevant documents.
```
Comment (mwest1066) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:47]: Add extra blank line before params.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 2]
```diff
@@ -3,20 +3,104 @@ import { OpenAI } from 'openai';
...
+ * @param client The OpenAI client to use.
+ * @param prompt The user's question generation prompt.
+ * @param authnUserId The user's authenticated user ID.
+ * @returns A string of all relevant context documents.
+ */
+async function makeContext(client: OpenAI, prompt: string, authnUserId): Promise<string> {
```
Comment (mwest1066) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:53]: `authnUserId` doesn't have a type

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 3]
```diff
@@ -3,20 +3,104 @@ import { OpenAI } from 'openai';
...
+  const context = contextDocs.join('\n\n');
+  return context;
+}
+
+/**
+ * Extracts the generated HTML and Python code from an OpenAI completion into job parameters.
+ * @param completion The completion to extract from.
+ * @param job The job whose data we want to extract into.
+ */
+function extractFromCompletion(completion, job: ServerJob): void {
```
Comment (mwest1066) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:72]: Another missing type.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 4]
```diff
@@ -3,20 +3,104 @@ import { OpenAI } from 'openai';
...
+    QuestionGenerationContextEmbeddingSchema,
+  );
+
+  const contextDocs = docs.map((doc) => doc.doc_text);
+  const context = contextDocs.join('\n\n');
+  return context;
+}
+
+/**
+ * Extracts the generated HTML and Python code from an OpenAI completion into job parameters.
```
Comment (mwest1066) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:68]: Blank line needed.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 5]
```diff
@@ -87,14 +139,98 @@ Keep in mind you are not just generating an example; you are generating an actua
...
+ */
+export async function regenerateQuestion(
+  client: OpenAI,
+  courseId: string | undefined,
+  authnUserId: string,
+  originalPrompt: string,
+  revisionPrompt: string,
+  originalHTML: string,
+  originalPython: string,
+) {
```
Comment (mwest1066) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:171]: Should specify return type.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 6]
```diff
@@ -87,14 +139,98 @@ Keep in mind you are not just generating an example; you are generating an actua
...
-    // TODO: Process generated documents
+    extractFromCompletion(completion, job);
+    job.data['prompt'] = revisionPrompt;
+    job.data['originalPrompt'] = originalPrompt;
+    job.data['generation'] = completion.choices[0].message.content;
+    job.data['context'] = context;
   });
 
-  return serverJob.jobSequenceId;
+  return { jobSequenceId: serverJob.jobSequenceId, completionPromise: jobPromise };
```
Comment (mwest1066) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:235]: Why are we returning a completion promise rather than just awaiting that promise in this function?

File: apps/prairielearn/src/lib/editors.ts [hunk 7]
```diff
@@ -879,9 +882,39 @@ export class QuestionAddEditor extends Editor {
...
     const fromPath = path.join(EXAMPLE_COURSE_PATH, 'questions', 'demo', 'calculation');
     const toPath = questionPath;
+
     debug(`Copy template\n from ${fromPath}\n to ${toPath}`);
     await fs.copy(fromPath, toPath, { overwrite: false, errorOnExist: true });
 
+    if (this.files != null) {
+      debug('Remove template files when file texts provided');
+      await fs.remove(path.join(toPath, 'question.html'));
+      await fs.remove(path.join(toPath, 'server.py'));
```
Comment (mwest1066) [inline apps/prairielearn/src/lib/editors.ts:892]: This will unconditionally remove both `question.html` and `server.py` if any files are provided. Is this even needed, or will the `writeFile()` below replace these files anyway?

File: apps/prairielearn/src/lib/server-jobs.ts [hunk 8]
```diff
@@ -33,20 +33,24 @@ interface ServerJobExecOptions {
...
+}
+
 export interface ServerJob {
   fail(msg: string): never;
   error(msg: string): void;
   warn(msg: string): void;
   info(msg: string): void;
   verbose(msg: string): void;
-  exec(file: string, args?: string[], options?: ServerJobExecOptions): Promise<void>;
+  exec(file: string, args?: string[], options?: ServerJobExecOptions): Promise<ServerJobResult>;
```
Comment (mwest1066) [inline apps/prairielearn/src/lib/server-jobs.ts:46]: I'd like @nwalters512's eyes on this as a change to the core server jobs code.

File: apps/prairielearn/src/lib/editors.ts [hunk 9]
```diff
@@ -851,12 +851,15 @@ export class CourseInstanceAddEditor extends Editor {
 export class QuestionAddEditor extends Editor {
   public readonly uuid: string;
 
-  constructor(params: BaseEditorOptions) {
+  files?: Record<string, string>;
+
+  constructor(params: BaseEditorOptions & { files?: Record<string, string> }) {
```
Comment (mwest1066) [inline apps/prairielearn/src/lib/editors.ts:856]: Tagging this for attention from @nwalters512 as a change to the core editor code.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 1 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:47]: Done in 5af63ea5a9d690527f5feb43c89b6d6ea7c7ba30

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 2 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: Done in 5af63ea5a9d690527f5feb43c89b6d6ea7c7ba30

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 3 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: Done in 5af63ea5a9d690527f5feb43c89b6d6ea7c7ba30

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 4 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:69]: Done in 5af63ea5a9d690527f5feb43c89b6d6ea7c7ba30

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 5 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: Done in 5af63ea5a9d690527f5feb43c89b6d6ea7c7ba30

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 6 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: Switched to awaiting in function in 5af63ea5a9d690527f5feb43c89b6d6ea7c7ba30

File: apps/prairielearn/src/lib/editors.ts (hunk 7 above)
Comment (jlevine18) [reply apps/prairielearn/src/lib/editors.ts:892]: Since we're copying from the basic question template, both its `question.html` and `server.py` are copied over. If a question only needs html, then that `server.py` contains unneeded code and can be removed. (I don't think there's a case where no html makes a valid question, so `question.html` should be assumed mandatory if `server.py` is provided)

File: apps/prairielearn/src/lib/server-jobs.ts (hunk 8 above)
Comment (jlevine18) [reply apps/prairielearn/src/lib/server-jobs.ts:46]: Discussed this afternoon.

File: apps/prairielearn/src/lib/editors.ts (hunk 9 above)
Comment (jlevine18) [reply apps/prairielearn/src/lib/editors.ts:856]: Discussed this afternoon.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 10]
```diff
@@ -87,14 +144,100 @@ Keep in mind you are not just generating an example; you are generating an actua
...
-    // TODO: Process generated documents
+    extractFromCompletion(completion, job);
+    job.data['prompt'] = revisionPrompt;
+    job.data['originalPrompt'] = originalPrompt;
+    job.data['generation'] = completion.choices[0].message.content;
+    job.data['context'] = context;
   });
 
-  return serverJob.jobSequenceId;
+  const jobData = await jobPromise;
```
Comment (mwest1066) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:241]: It's a bit clunky to store the promise and then await it later. Instead you can remove this line and above just do:
```
const jobData = await serverJob.execute(async (job) => {
```

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 6 above)
Comment (mwest1066) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: See https://github.com/PrairieLearn/PrairieLearn/pull/10221#discussion_r1679724211

File: apps/prairielearn/src/lib/editors.ts (hunk 7 above)
Comment (mwest1066) [reply apps/prairielearn/src/lib/editors.ts:892]: Good point. Does this apply to other files as well? I guess I'm bothered by the fact that at the moment we single out these two files for special treatment. Right now this is fine, because these are the only two we are generating in any case, but this function is being written as a general mechanism for updating questions.

I think that we should change the function to delete all existing files in the question if replacement files are provided, and document that this is what will happen. The only exception to this should be `info.json`, I guess? How is the updating of that file handled in this flow?

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 6 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: Changed in b5c604cd1a9f33f625b695563a12dcab48f3d0f9

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 10 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: Changed in b5c604cd1a9f33f625b695563a12dcab48f3d0f9.

File: apps/prairielearn/src/lib/editors.ts (hunk 7 above)
Comment (jlevine18) [reply apps/prairielearn/src/lib/editors.ts:892]: Made a small update in b5c604cd1a9f33f625b695563a12dcab48f3d0f9- the new logic is that since the template has `question.html`, `server.py` and `info.json`, if _any_ files are provided, files _only_ exist if they are provided, _except_ for `info.json` which will default to the template file if not provided

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 11]
```diff
@@ -3,72 +3,129 @@ import { OpenAI } from 'openai';
...
+        
+        Parameters can be read in \`question.html\` with Mustache syntax. For instance, if \`server.py\` contains \`data["params"]["answer"]\`, it can be read with \`{{ params.answer }}\` in \`question.html\`.
+        
+        # Context
+        
+        Here is some context that may help you respond to the user. This context may include example questions, documentation, or other information that may be helpful.
+        
+        ${context}
+        
+        `;
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:43]: Nit: dedent all of this? A backtick string literal includes all internal whitespace, including that at the start of every line here.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 12]
```diff
@@ -3,72 +3,129 @@ import { OpenAI } from 'openai';
...
+
+  const docs = await queryRows(
+    sql.select_nearby_documents,
+    { embedding: vectorToString(embedding), limit: 5 },
+    QuestionGenerationContextEmbeddingSchema,
+  );
+
+  const contextDocs = docs.map((doc) => doc.doc_text);
+  const context = contextDocs.join('\n\n');
+  return context;
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:65]: Nit, a little more concise:

```suggestion
  return docs.map((doc) => doc.doc_text).join('\n\n');
```

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 13]
```diff
@@ -87,14 +144,98 @@ Keep in mind you are not just generating an example; you are generating an actua
       user: openAiUserFromAuthn(authnUserId),
     });
 
-    const completionText = completion.choices[0].message.content;
+    extractFromCompletion(completion, job);
+    job.data['prompt'] = prompt;
+    job.data['generation'] = completion.choices[0].message.content;
+    job.data['context'] = context;
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:150]: I might consider including the entire `completion` in `job.data`, both for debugging and so that we can store + retrieve usage information in a structured way.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 14]
```diff
@@ -87,14 +144,98 @@ Keep in mind you are not just generating an example; you are generating an actua
...
+    type: 'llm_question_regen',
+    description: 'Revise a question using the LLM',
+  });
+
+  const jobData = await serverJob.execute(async (job) => {
+    job.info(`prompt is ${revisionPrompt}`);
+
+    const context = await makeContext(client, originalPrompt, authnUserId);
+    const sysPrompt = `
+        ${promptPreamble(context)}
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:188]: Same comment as above, consider dedenting this.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 15]
```diff
@@ -1,18 +1,66 @@
...
-import { generateQuestion } from '../../lib/aiQuestionGeneration.js';
+import { idsEqual } from '../../../lib/id.js';
+import { selectJobsByJobSequenceId } from '../../../lib/server-jobs.js';
+import { generateQuestion, regenerateQuestion } from '../../lib/aiQuestionGeneration.js';
 import { syncContextDocuments } from '../../lib/contextEmbeddings.js';
 
-import { AiGeneratePage } from './instructorAiGenerateQuestion.html.js';
+import { AiGeneratePage, GenerationResults } from './instructorAiGenerateQuestion.html.js';
+
+type IdType = z.infer<typeof IdSchema>;
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:20]: No need for this, you can use `string` directly.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 16]
```diff
@@ -52,22 +100,84 @@ router.post(
     if (!config.openAiApiKey || !config.openAiOrganization) {
       throw new error.HttpStatusError(403, 'Not implemented (feature not available)');
     }
+
     const client = new OpenAI({
       apiKey: config.openAiApiKey,
       organization: config.openAiOrganization,
     });
 
-    if (req.body.__action === 'generate_question') {
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:60]: Why the change in action name?

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 17]
```diff
@@ -52,22 +100,84 @@ router.post(
...
-      const jobSequenceId = await generateQuestion(
+    if (req.body.__action === 'ai_generate') {
+      const jobSeq = await generateQuestion(
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:110]: I'd call this something like `result`. This is not a job sequence.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 18]
```diff
@@ -52,22 +100,84 @@ router.post(
...
+    } else if (req.body.__action === 'ai_regen') {
+      const genJobs = await selectJobsByJobSequenceId(req.body.unsafe_sequence_job_id);
+      if (
+        genJobs.length !== 1 ||
+        !genJobs[0]?.course_id ||
+        !idsEqual(genJobs[0]?.course_id, res.locals.course.id)
+      ) {
+        throw new error.HttpStatusError(
+          403,
+          `Course for job sequence ${req.body.unsafe_sequence_job_id} does not match the current course.`,
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:137]: We can just say something generic like "Job sequence #### not found".

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 19]
```diff
@@ -87,14 +144,98 @@ Keep in mind you are not just generating an example; you are generating an actua
...
+        # Prompt
+        
+        A user will now request your help in in revising the question that you generated. Respond in a friendly but concise way. Include \`question.html\` and \`server.py\` in Markdown code fences in your response, and tag each code fence with the language (either \`html\` or \`python\`). Omit \`server.py\` if the question does not require it (for instance, if the question does not require randomization).
+        
+        Keep in mind you are not just generating an example; you are generating an actual question that the user will use directly.
+        `;
+
+    job.info(`system prompt is: ${sysPrompt}`);
+
+    //TODO [very important]: normalize to prevent prompt injection attacks
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:222]: Micro-nit:

```suggestion
    // TODO [very important]: normalize to prevent prompt injection attacks
```

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 20]
```diff
@@ -52,22 +100,84 @@ router.post(
...
+
+      const qid = await saveGeneratedQuestion(
+        res,
+        genJobs[0]?.data?.html,
+        genJobs[0]?.data?.python,
+      );
+
+      if (qid) {
+        res.redirect(res.locals.urlPrefix + '/question/' + qid + '/settings');
+      }
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:180]: What happens if `qid` isn't defined? The request will hang. We should handle this case somehow.

File: apps/prairielearn/src/lib/editors.ts [hunk 21]
```diff
@@ -879,9 +882,43 @@ export class QuestionAddEditor extends Editor {
...
     debug(`Copy template\n from ${fromPath}\n to ${toPath}`);
     await fs.copy(fromPath, toPath, { overwrite: false, errorOnExist: true });
 
+    if (this.files != null) {
+      debug('Remove template files when file texts provided');
+      await fs.remove(path.join(toPath, 'question.html'));
+      await fs.remove(path.join(toPath, 'server.py'));
+
+      if ('info.json' in this.files) {
+        await fs.remove(path.join(toPath, 'info.json'));
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/editors.ts:895]: Not to explode the scope of this PR, but: AFAICT, this will always create a question with the tags `mwest` and `fa17`, which is pretty undesirable. cc @mwest1066, if we don't address that here, we should address it somewhere.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 22]
```diff
@@ -52,22 +100,84 @@ router.post(
...
+        req.body.prompt,
+        genJobs[0]?.data?.html,
+        genJobs[0]?.data?.python,
+      );
+
+      const data = jobSeq.jobData;
+      if (data) {
+        res.send(
+          GenerationResults(data.data.html, data.data.python, jobSeq.jobSequenceId, res.locals),
+        );
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:155]: There seems to be some terminology confusion here. Why does `jobData` itself have a `data` property? I'm assuming that `jobData` isn't named correctly.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 23]
```diff
@@ -52,22 +100,84 @@ router.post(
...
       );
-      res.redirect(res.locals.urlPrefix + '/jobSequence/' + jobSequenceId);
+
+      const data = jobSeq.jobData;
+      if (data) {
+        res.send(
+          GenerationResults(data.data.html, data.data.python, jobSeq.jobSequenceId, res.locals),
+        );
+      } else {
+        throw new error.HttpStatusError(500, `Job Sequence ${jobSeq.jobSequenceId} failed.`);
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:123]: Nit:

```suggestion
        throw new error.HttpStatusError(500, `Job sequence ${jobSeq.jobSequenceId} failed.`);
```

File: apps/prairielearn/src/lib/editors.ts (hunk 21 above)
Comment (mwest1066) [reply apps/prairielearn/src/lib/editors.ts:895]: Ugh, yes, we definitely want to stop doing this. @jlevine18 can you make an issue to fix this (outside the scope of this PR)?

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 11 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: done in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 13 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts:148]: done in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts (hunk 14 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: done in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts (hunk 15 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts]: done in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts (hunk 16 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:60]: reverted action naming in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts (hunk 17 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts]: changed in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts (hunk 18 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts]: changed in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts (hunk 20 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts]: it actually wouldn't because the qid-generating function would redirect, but that was too confusing and bad code anyways. changed to something more reasonable in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts (hunk 22 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts]: changed in e0331beaf9d587990dcda087d800feedc786fc87.

File: apps/prairielearn/src/lib/editors.ts (hunk 21 above)
Comment (jlevine18) [reply apps/prairielearn/src/lib/editors.ts:895]: See #10243.

Comment (nwalters512) [reply apps/prairielearn/src/lib/editors.ts:895]: Thanks! Resolving this for now.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 24]
```diff
@@ -52,22 +96,86 @@ router.post(
...
-      res.redirect(res.locals.urlPrefix + '/jobSequence/' + jobSequenceId);
+
+      const data = result.jobResult;
+      if (data) {
+        res.send(
+          GenerationResults(data.data.html, data.data.python, result.jobSequenceId, res.locals),
+        );
+      } else {
+        throw new error.HttpStatusError(500, `Job sequence ${result.jobSequenceId} failed.`);
+      }
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:120]: Nit, `data` is not a good alias for `jobResult`:

```suggestion
      const data = result.jobResult;
      if (result.jobResult) {
        res.send(
          GenerationResults(result.jobResult.data.html, result.jobResult.data.python, result.jobSequenceId, res.locals),
        );
      } else {
        throw new error.HttpStatusError(500, `Job sequence ${result.jobSequenceId} failed.`);
      }
```

You might consider having `generateQuestion` return structured data instead of the entire `jobResult` object. That is, it can return `{ jobSequenceId, htmlFileContents, pythonFileContents, ... }`.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 25]
```diff
@@ -52,22 +96,86 @@ router.post(
...
+      );
+
+      const data = result.jobResult;
+      if (data) {
+        res.send(
+          GenerationResults(data.data.html, data.data.python, result.jobSequenceId, res.locals),
+        );
+      } else {
+        res.redirect('/pl/jobSequence/' + result.jobSequenceId);
+      }
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:154]: Same comment as above, if you're so inclined.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts [hunk 26]
```diff
@@ -3,16 +3,60 @@ import asyncHandler from 'express-async-handler';
...
+    locals: res.locals,
+    files,
+  });
+
+  const serverJob = await editor.prepareServerJob();
+
+  try {
+    await editor.executeWithServerJob(serverJob);
+  } catch (err) {
+    return { redirect: res.locals.urlPrefix + '/edit_error/' + serverJob.jobSequenceId };
```
Comment (nwalters512) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts:48]: Another option for control flow here: we have support for "throwing" redirects, like you can see here:

https://github.com/PrairieLearn/PrairieLearn/blob/ee7b9af43448e50ffbdd221bcca595c90d80d19a/apps/prairielearn/src/ee/lib/terms.ts#L55

This can keep the caller oblivious of the failure state here, which can be marginally cleaner.

Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts]: Changed in 6919d9f4829f12ab87c64d40f2dccf405a98d563.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts (hunk 25 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts]: Changed in 6919d9f4829f12ab87c64d40f2dccf405a98d563.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts (hunk 24 above)
Comment (jlevine18) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts]: Changed in 6919d9f4829f12ab87c64d40f2dccf405a98d563.

**Review by nwalters512 — APPROVED** [review]
